src/backend/tcop/postgres.c

```diff
@@ -5330,6 +5330,13 @@ PostgresMain(int argc, char *argv[],
 						ereport(ERROR,
 								(errcode(ERRCODE_PROTOCOL_VIOLATION),
 								 errmsg("MPP protocol messages are only supported in QD - QE connections")));
+
+#ifdef FAULT_INJECTOR
+					if (SIMPLE_FAULT_INJECTOR("vmem_oom_set_tracked_bytes") == FaultInjectorTypeSkip)
+					{
+						VmemTracker_SetTrackedBytes(VmemTracker_GetReservedVmemBytes()+(1<<VmemTracker_GetChunkSizeInBits())-20);
```

Comment:
Could you please explain more about this formula?

Besides, you may want to use `CHUNKS_TO_BYTES()` ?
```suggestion
						VmemTracker_SetTrackedBytes(VmemTracker_GetReservedVmemBytes()+CHUNKS_TO_BYTES(1)-20);
```

Reply:
Under the circumstance that session 1  has consumed all vmem chunks including waivedChunks, we set session 2's trackedBytes close to threshold that it needs to reserve a new chunk, so that it can easily trigger `gp_malloc` and find that no available chunks remained. Then session 2 will the handle OOM error messages, and call `gp_malloc` again.

Reply:
Thanks for your explanation. I'm curious that why we need to substract 20-byte from it? What does the 20-byte stand for?

Reply:
HAHA, 20 bytes is kind of a small number I chosen arbitrarily, so that it can easily trigger the logic of reserving new vmem chunks when `gp_malloc` is called. Of course, other small number is also OK, like 100, 200 and so on.

Reply:
> CHUNKS_TO_BYTES

Thanks for the comment, the reason that I didn't use `CHUNKS_TO_BYTES(1)` is that it's a macro defined in `vmem_tracker.c`, so we can't use it in `postgres.c`.

Reply:
> this fix ensures that every QE process has it own waived vmem chunks to handle error messages.



> Thanks for your explanation. I'm curious that why we need to substract 20-byte from it? What does the 20-byte stand for?

I go through the related code again, and found there is no need to subtract a small number given that the trackedBytes has been set to `VmemTracker_GetReservedVmemBytes()+(1<<VmemTracker_GetChunkSizeInBits())`. I modified it in the latest push.